File: 965/965.py
# @brief:  leetcode-965-单值二叉树
# @envir:  python 2.7
# @date:   2019/5/7
# @author: luhao

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.num = 0
        self.first = True
        
    def isUnivalTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        if self.first:
            self.num = root.val
            self.first = False
        else:
            if self.num != root.val:
                return False
        return self.isUnivalTree(root.left) and self.isUnivalTree(root.right)

File: 965/965.cpp
/*!***********************
* @brief:  leetcode-965-单值二叉树
* @date:   2019/5/8
* @envir:  gcc 7.4.0
* @author: luhao
* ***********************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int first = true;
    int num;
    bool isUnivalTree(TreeNode* root) {
        if (!root)
            return 1;
        if (first) {
            num = root->val;
            first =  false;
        }
        else {
            if (root->val != num)
                return 0;
        }
        return isUnivalTree(root->left) && isUnivalTree(root->right);
    }
};
